climada/util/coordinates.py

```diff
@@ -76,6 +78,273 @@
+
+    # query the k closest points of the n_points using dual tree
+    dist, assigned = tree.query(np.radians(coordinates[idx]), k=1,
+                                return_distance=True, dualtree=True,
+                                breadth_first=False)
+
+    # Raise a warning if the minimum distance is greater than the
+    # threshold and set an unvalid index -1
+    num_warn = np.sum(dist * EARTH_RADIUS_KM > threshold)
+    if num_warn:
+        LOGGER.warning('Distance to closest centroid is greater than %s'
+                       'km for %s coordinates.', threshold, num_warn)
+        assigned[dist * EARTH_RADIUS_KM > threshold] = -1
+
+    # Copy result to all exposures and return value
+    return np.squeeze(assigned[inv])
+
+def interpolate_lines(gdf_lines, point_dist=5):
+    """ Convert a GeoDataframe with LineString geometries to 
+    Point geometries, where Points are placed at a specified distance along the
+    original LineString 
+    Important remark: LineString.interpolate() used here performs interpolation 
+    in a Cartesian coordinate system. The result will be inaccurate for 
+    LineStrings whose defining points are spaced very far from each other 
+    (like end points in Zürich and NYC)
+    
+    Parameters
+    ----------
+    gdf_lines : gpd.GeoDataframe or str
+        Geodataframe or filepath from which to read the GeoDataframe
+    point_dist : float
+    Distance in metres apart from which the generated  Points should be placed.
+    
+    Returns
+    -------
+    gdf_points : gpd.GeoDataFrame
+        with individual Point per row, retaining all other column infos
+        belonging to its corresponding line (incl. line length of original geom.
+        and multi-index referring to original indexing)
+        
+    See also
+    --------
+    * util.coordinates.compute_geodesic_lengths()
+    * util.lines_polys_handler.point_exposure_from_lines()
+    """
+
+    if not isinstance(gdf_lines, gpd.GeoDataFrame):
+        gdf_lines = gpd.read_file(gdf_lines)
+        
+    gdf_points = gdf_lines.copy()
+    gdf_points['length'] = np.ones(len(gdf_points))*point_dist
+    gdf_points['length_full'] = compute_geodesic_lengths(gdf_points)
+
+    # split line lengths into relative fractions acc to point_dist (e.g. 0, 0.5, 1)
+    gdf_points['distance_vector'] = gdf_points.apply(
+        lambda row: np.linspace(0, 1, num=int(np.ceil(row.length_full/
+                                                      row.length)+1)), axis=1)
+
+    # create MultiPoints along the line for every position in distance_vector
+    gdf_points['geometry'] = gdf_points.apply(
+        lambda row: MultiPoint(
+            [row.geometry.interpolate(dist, normalized=True) 
+             for dist in row.distance_vector]), axis=1)
+    
+    # expand gdf from MultiPoint entries to single Points per row
+    return gdf_points.explode().drop(['distance_vector', 'length_full'], axis=1)
+
+def interpolate_polygons(gdf_poly, area_point):
```

Comment:
Would it be useful to provide the grid to map the polygons onto? Think for instance that one has a centroids raster and wants to map the polygons to this raster.

Reply:
ehm. if you already have a grid at hand, you may not really need this function. with a grid, you can just use an sjoin to see which points of the grid lie inside the polygon. The idea of this func here is to help the user make a grid, where they just need to think about the representative area (resolution) per grid-point.

Reply:
Ok, but then the function should only produce the grid. Another function can then be made that maps the polygons onto said grid.

Reply:
Note that this function doesn't simply create "a grid". It creates a separate grid for each polygon. Not only the origin of each of those grids is different, but also the resolution. The reason for that is the fact that polygons are assumed to be given in lat-lon-coordinates, while the user specifies the desired resolution in meters. So, the function chooses the grid resolution according to the specific degree-meter conversion factor at the latitudinal location of each polygon.

I partly agree with @chahank: This function should probably be better separated into "subfunctions". I think the most important separation would be to have a function `_interpolate_one_polygon` that works on each polygon separately and does both generate the grid and apply it to the polygon. The function `interpolate_polygons` then just uses `apply` and `explode`. If you want to be even more modular (as suggested by @chahank), you can divide `_interpolate_one_polygon` further.

Reply:
Both good points! I will implement a more modular version and also think about your use case @chahank , which was not the prime intention, and very easily implementable with the default geopandas functionalities. It might still fit somewhere here.

Reply:
Great! I think this is really important for your design and should be taken into account right at the beginning. Even though it might not be the prime intention, a well-designed set of functions should without difficulty allow for both use cases. Whether this is possible might be a good way to test the (good) modularity of the design.

I think the most simple use-case is something like LitPop where a given value is distributed over all chosen points inside of a shape-polygon. Afterwards, there should be a function where the impacts are aggregated.

Now, the variations and difficulties come in when one chooses a way to distribute, a way to set the points inside of the polygon (a grid, centroids, grid of equal surface cells,...) and a way to aggregate the impacts.

Reply:
commit a4ceae19 refactors the polygon (and the line) interpolation: a sub-function `_interpolate_one_polygon` now creates a separate grid for each polygon with the desired resolution and cuts out only those resulting points that lie inside it.